File: test-src/26_nullSafety.kt
// See https://kotlinlang.org/docs/reference/null-safety.html

class Department {
  var head: User? = null
}

class User(var name: String? = null) {
  var department: Department? = null
}

fun main(args: Array<String>) {
  var a: String = "abc"
  //a = null // compilation error

  var b: String? = "abc"
  b = null // ok

  println("a.length = ${a.length} (should be 3)")

  //val blInvalid = b.length // error: variable 'b' can be null

  // SWIFT: let bl = b != nil ? b!.length : -1
  val bl = if (b != null) b.length else -1
  println("bl = $bl (should be -1)")

  // SWIFT: if b != nil && b!.length > 0 {
  if (b != null && b.length > 0) {
    // SWIFT: print("ERROR: String of length \(b!.length)")
    println("ERROR: String of length ${b.length}")
  } else {
    println("OK: Empty string")
  }

  println("${b?.length} should be null")

  // Classes
  val bob = User(name = "Bob")
  val john = User(name = "John")
  val marketing = Department()
  bob.department = marketing
  marketing.head = john
  println("${bob.department?.head?.name} should be John")

  b = "new b value"
  println("${b!!.length} should be 11")

  val aInt: Int? = a as? Int
  println("$aInt should be null")
}

File: test-dest/26_nullSafety.swift
// See https://kotlinlang.org/docs/reference/nil-safety.html

public class Department {
  var head: User? = nil
  init() {
  }
}

public class User {
  var name: String?

  var department: Department? = nil
  init(name: String? = nil) {
    self.name = name
  }
}

func main(args: [String]) {
  var a: String = "abc"
  //a = nil // compilation error

  var b: String? = "abc"
  b = nil // ok

  print("a.length: \(a.length) (should be 3)")

  //let blInvalid = b.length // error: variable 'b' can be nil

  let bl = b != nil ? b!.length : -1
  print("bl: \(bl) (should be -1)")

  if b != nil && b!.length > 0 {
    print("ERROR: String of length \(b!.length)")
  } else {
    print("OK: Empty string")
  }

  print("\(b?.length) should be nil")

  // Classes
  let bob = User(name: "Bob")
  let john = User(name: "John")
  let marketing = Department()
  bob.department = marketing
  marketing.head = john
  print("\(bob.department?.head?.name) should be John")

  b = "new b value"
  print("\(b!.length) should be 11")

  let aInt: Int32? = a as? Int32
  print("\(aInt) should be nil")
}

main([])
